tests/kernel/timer/timer_behavior/src/main.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+
+#include <zephyr/zephyr.h>
+
+#include <tc_util.h>
+#include <ztest.h>
+
+static uint32_t periodic_idx = 0;
+static uint64_t periodic_data[CONFIG_TIMER_TEST_SAMPLES + 1];
+static uint64_t periodic_start, periodic_end;
+static struct k_timer periodic_timer;
+K_SEM_DEFINE(periodic_sem, 0, 1);
+
+void periodic_fn(struct k_timer *t)
+{
+	if (periodic_idx < CONFIG_TIMER_TEST_SAMPLES) {
+
+#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+		periodic_end = k_cycle_get_64();
+#else
+		periodic_end = k_cycle_get_32();
+#endif
+		periodic_data[periodic_idx + 1] = periodic_end;
+		periodic_idx++;
+	} else {
+		k_timer_stop(t);
+		k_sem_give(&periodic_sem);
+	}
+}
+
+ZTEST(timer_behavior, periodic_behavior)
+{
+	TC_PRINT("periodic timer behavior test\n");
+
+	k_timer_init(&periodic_timer, periodic_fn, NULL);
+
+#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	periodic_start = k_cycle_get_64();
+#else
+	periodic_start = k_cycle_get_32();
+#endif
+
+	periodic_data[0] = periodic_start;
+	k_timer_start(&periodic_timer, K_NO_WAIT, K_USEC(CONFIG_TIMER_TEST_PERIOD));
+	k_sem_take(&periodic_sem, K_FOREVER);
+
+	TC_PRINT("periodic timer samples gathered, calculating statistics\n");
+
+	/* calculate variance, and precision */
+	uint64_t total_cycles = 0;
+
+	for (int i = 0; i < CONFIG_TIMER_TEST_SAMPLES; i++) {
+		total_cycles += (periodic_data[i + 1] - periodic_data[i]);
+	}
+
+	double mean = (double)total_cycles / (double)CONFIG_TIMER_TEST_SAMPLES;
+
+	double variance = 0;
+
+	for (int i = 0; i < CONFIG_TIMER_TEST_SAMPLES; i++) {
+		double diff = (double)(periodic_data[i + 1] - periodic_data[i]) - mean;
```

Comment:
This doesn't handle wraparound of non-64-bit cycle counts correctly as it's doing the math in 64 bit precision (i.e. you want `0x00001000 - 0xfffff000` to produce 0x2000, but it actually gives you a huge 64 bit value instead).  My suggestion would be do do the interval subtraction in the k_timer callback instead, which can be guaranteed not to overflow as long as the interval isn't pathologically long.

Also, note that ARM SysTick is actually a **24 bit** hardware counter, so k_cycle_get_32() wraps at like 6 Hz on  the faster boards (There was an Atmel board that was susceptible to this, forget the name).

Reply:
That is really helpful to know, thank you!

I've dropped all the rollover differences in the stats, but without knowing the maximum count of the timer (is this Kconfig/dts defined somewhere? if not that would be nice) I don't think knowing the true total time elapsed will work in all cases. I'll try it on some more hardware next week.